src/main/java/pl/damianszczepanik/jenkins/buildhistorymanager/model/actions/DeleteArtifactsWithPatternAction.java

```diff
@@ -0,0 +1,175 @@
+
+/**
+ * Deletes the artifacts with patterns.
+ */
+public class DeleteArtifactsWithPatternAction extends Action {
+    private static final Logger LOGGER = Logger.getLogger(DeleteArtifactsWithPatternAction.class.getName());
+
+    private String include;
+    private String exclude;
+
+    public String getInclude() {
+        return include;
+    }
+
+    @DataBoundSetter
+    public void setInclude(String include) {
+        this.include = include;
+    }
+
+    public String getExclude() {
+        return exclude;
+    }
+
+    @DataBoundSetter
+    public void setExclude(String exclude) {
+        this.exclude = exclude;
+    }
+
+    @DataBoundConstructor
+    public DeleteArtifactsWithPatternAction() {
+        // Jenkins stapler requires to have public constructor with @DataBoundConstructor
+    }
+
+    public static String removeLastSlash(String archiveRootPath) {
+        if(archiveRootPath.endsWith("/")) {
+            return archiveRootPath.substring(0, archiveRootPath.lastIndexOf("/"));
+        } else {
+            return archiveRootPath;
+        }
+    }
+
+    public static boolean isDirEmpty(Path path) throws IOException {
+        if (Files.isDirectory(path)) {
+            try (Stream<Path> entries = Files.list(path)) {
+                return !entries.findFirst().isPresent();
+            }
+        }        
+        return false;
+    }
+
+    // if 'file' is on a different node, this FileCallable will be transferred to that node and executed there.
+    public static final class Delete implements FileCallable<Void> {
+        private static final long serialVersionUID = 1;
+        private final String archiveRootPath;
+
+        public Delete(String archiveRootPath) {
+            this.archiveRootPath = removeLastSlash(archiveRootPath);
+        }
+
+        @Override 
+        public Void invoke(File vFile, VirtualChannel channel) throws IOException {          
+
+            Set<File> directories = new HashSet<>();
+            deleteFileOrLogError(vFile, directories);
+            deleteEmptyDirectoriesAndParents(directories);
+            return null;
+        }
+
+        public void deleteEmptyDirectoriesAndParents(Set<File> directories) throws IOException {
+            for (File dir : directories){
+                if (shouldDeleteDirectory(dir)) {
+                    Util.deleteFile(dir);
+                    deleteParentDirectories(dir.getParentFile());
+                }
+            }
+        }
+
+        public boolean shouldDeleteDirectory(File dir) throws IOException {
+            if (isInvalidDirectory(dir)) {
+                return false;
+            }
+
+            File parent = dir.getParentFile();
+            return hasValidParent(parent);
+        }
+
+        public boolean hasValidParent(File parent) {
+            return parent != null && parent.getPath().equals(this.archiveRootPath);
+        }
+
+        public boolean isInvalidDirectory(File dir) throws IOException {
```

Comment:
avoid using negation in method name

Reply:
Renamed `isInvalidDirectory` to 'isValidDirectory' and updated the unit tests.